File: catkin_ws/src/navigation_tutorial/src/avoidance.cpp
#include <vector>

#include "ros/ros.h"
#include "geometry_msgs/Twist.h"
#include "sensor_msgs/LaserScan.h"

sensor_msgs::LaserScan scan;

void scan_callback(const sensor_msgs::LaserScan::ConstPtr& msg)
{
    scan = *msg;
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "go_straight");

    ros::NodeHandle nh;

    ros::Publisher vel_pub = nh.advertise<geometry_msgs::Twist>("/cmd_vel", 1);

    ros::Subscriber scan_sub = nh.subscribe("/scan", 1, scan_callback);

    ros::Rate loop_rate(10);

    while (ros::ok()){
        geometry_msgs::Twist vel;
        if(!scan.ranges.empty()){
            std::vector<float>::iterator min_range = std::min_element(
                    scan.ranges.begin()+int(scan.ranges.size()*0.25),
                    scan.ranges.end()-int(scan.ranges.size()*0.25));
            size_t min_idx = std::distance(scan.ranges.begin(), min_range);
            ROS_INFO_STREAM("min_range: "<< scan.ranges[min_idx] << ", min_idx: " << min_idx);
            if(scan.ranges[min_idx] >= 0.4){
                vel.linear.x = 0.3;
                vel.angular.z = 0.0;
            }else{
                vel.linear.x = 0.0;
                if(min_idx < scan.ranges.size()/2.0){
                    vel.angular.z = 0.5;
                }else{
                    vel.angular.z = -0.5;
                }

            }
        }
        vel_pub.publish(vel);
        ROS_INFO_STREAM("vel: "<< vel);
        ros::spinOnce();
        loop_rate.sleep();
    }

    return 0;
}

File: catkin_ws/src/navigation_tutorial/scripts/avoidance.py
#!/usr/bin/env python3
import numpy as np

import rospy
from geometry_msgs.msg import Twist
from sensor_msgs.msg import LaserScan

class Avoidance:
    def __init__(self):
        rospy.init_node('avoidance', anonymous=True)
        
        # Publisher
        self.cmd_vel_pub = rospy.Publisher('/planner/cmd_vel', Twist, queue_size=10)

        # Subscriber
        scan_sub = rospy.Subscriber('/scan', LaserScan, self.callback_scan)

        self.min_range = None

    def callback_scan(self, data):
        fov = np.deg2rad(60)
        min_range = data.range_max
        min_idx = -1
        angle = data.angle_min
        for idx, r in enumerate(data.ranges):
            angle += data.angle_increment
            if -fov<angle<fov:
                if r<min_range:
                    min_range = r
                    min_idx = idx
        if min_idx < len(data.ranges)/2.0:
            self.direction = "RIGHT"
        else:
            self.direction = "LEFT"
        self.min_range = min_range

    def process(self):
        r = rospy.Rate(10)
        while not rospy.is_shutdown():
            vel = Twist()
            if self.min_range is not None:
                if self.min_range >= 0.4:
                    vel.linear.x = 0.2
                    vel.angular.z = 0.0
                else:
                    vel.linear.x = 0.0
                    if self.direction == "RIGHT":
                        vel.angular.z = 0.5
                    elif self.direction == "LEFT":
                        vel.angular.z = -0.5
            self.cmd_vel_pub.publish(vel)
            r.sleep()

if __name__=='__main__':
    avoidance = Avoidance()
    try:
        avoidance.process()
    except rospy.ROSInitException:
        pass
